### DISCUSSION TITLE
How do dynamic redirect?

### DISCUSSION THREAD

**Discussion by pquerner** [pleerock/routing-controllers#1329] [Q&A]
Hello!

I have a route which create a model. If checks succeeds and the model could be saved to the database, I want to render a template.
If it fails, I want it to render a different template.

Like so:

```
@Controller('/pn')
export class PushNotificationsController extends BaseController {

@Post('/add')
    async createdOrError(@Res() response: Response, @Req() request: Request, @Body() pn: PushNotificationCreationAttributes) {
// ..

try {
            // save model
        } catch (error) {
            //todo: render error
            console.error(error);
            response.render('page/error.njk', {...});
            return;
        }
        return response.render('page/pn/created.njk', {...});

}
``` 

I have tried `response.render(...) return;`. but also direct `return response render(...)`.

Getting error 
```
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client

```

**Comment by attilaorosz** [comment]
According to the docs, the render method does not return anything, it just sends the response internally. The problem with this is that routing-controllers will try to return whatever your controller returns, resulting in a 'double response'. However, if you return the response object itself, routing-controller will not try to do any additional response writing. Please try 
```typescript
response.render(...);
return response;
```

**Reply by pquerner** [reply]
Hello!

```
@Get('/test')
    async test(@Res() response: Response, @Req() request: Request) {
        console.log("test");
        if(true) {
            console.log("test2");
            response.render('page/empty.njk');
            return response;
        }
        console.log("test3");

        response.render('page/register.njk');
        return response;
    }
```

This simple route outputs the "empty" template, but then dies at 

```
node:_http_outgoing:782
    throw new ERR_HTTP_HEADERS_SENT('remove');
    ^
```

Full stacktrace:

```
Error [ERR_HTTP_HEADERS_SENT]: Cannot remove headers after they are sent to the client
    at ServerResponse.removeHeader (node:_http_outgoing:782:11)
    at write (/Users/p.querner/projects/x/node_modules/finalhandler/index.js:282:9)
    at AsyncResource.runInAsyncScope (node:async_hooks:206:9)
    at listener (/Users/p.querner/projects/x/node_modules/on-finished/index.js:170:15)
    at onFinish (/Users/p.querner/projects/x/node_modules/on-finished/index.js:101:5)
    at callback (/Users/p.querner/projects/x/node_modules/ee-first/index.js:55:10)
    at IncomingMessage.onevent (/Users/p.querner/projects/x/node_modules/ee-first/index.js:93:5)
    at IncomingMessage.emit (node:events:526:35)
    at IncomingMessage.emit (node:domain:488:12)
    at endReadableNT (node:internal/streams/readable:1589:12) {
  code: 'ERR_HTTP_HEADERS_SENT'
}

```

a page reload is refused, since the listener completely died.

The same result is when I change `if (true)..` to `if (false)..`. The "register" template is rendered and then the same behaviour is observed.

`async` isnt used here, but removing it yields the same results in all cases.

**Reply by attilaorosz** [reply]
Do you happen to have any middleware that runs after the controller? Or do you have a route that also matches your target route? If the response object is returned the .next method is called on response.
Another possible cause could be that you write the response in a before middleware and thats why the render fails.

**Reply by pquerner** [reply]
I use a middleware for session handling, and to resolve files from a "public" directory.

```
...use(express.static('public'));
...use(
            session({
                // ...
            })
        )
```

In the controller file no other routes are defined (that GET /test is all there is), nor any other middlewares.

The full controller file looks like that:

```
import {Controller, Get, Req, Res} from "routing-controllers";
import {Request, Response} from "express";

@Controller('/test')
export class TestController {
    @Get('/test')
    test(@Res() response: Response, @Req() request: Request) {
        console.log("test");
        if(false) {
            console.log("test2");
            response.render('page/empty.njk');
            return response;
        }
        console.log("test3");

        response.render('page/register.njk');
        return response;
    }
}

```

**Reply by attilaorosz** [reply]
Could you create a repro repo so I can try to debug this? According to the stacktrace, the error is originating from the default express error handler.

**Reply by pquerner** [reply]
You're right. I've setup a bare-bones clean repo with and it works there just fine.
I'll keep digging and update here when I find something.

**Reply by pquerner** [reply]
So far I havent found anything, but I added empty middleware and the error went away:

```
        useExpressServer(App.expressInstance, {
            controllers: [path.join(__dirname + '/controllers/*Controller.js')],
        });
        App.expressInstance.use(function(req, res, next) { // <-- without this the "ERR_HTTP_HEADERS_SENT" happens (Oô)
            next();
        });
```
